Implement configuraiton for typing unloaded linked resources

furinvader opened the pull request:
When a linked resource, like i.e. `workPackage.parent` was not embedded, it would always have `HalResource` as its type, despite it being a `WorkPackageResource`.

This PR allows to configure the class of linked and unloaded resources.

**Todo**
- [x] Make the `className` argument of the configuration optional - We do not want to create classes for each type to type its linked resources.
- [x] Correctly apply getters and setters after the resource has been loaded

furinvader committed:
Refactor halTransformConfig

Use the opApiModule.

furinvader committed:
Refactor HalResource

Get rid of the redundant _$embedded and _$links
properties.

furinvader committed:
Refactor the halTransform components

furinvader committed:
Fix HalLink service tests

furinvader committed:
Implement resource type configuration for unloaded resources

This change makes it possible to correctly type linked resources,
which would not inherit from their respective resource class, if
not already embedded.

The HalResource class was extended by a factory method,
rendering the halTransform service useless.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://furinvader-feature-typing-unloaded-linked-resources.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/' target='_blank'>Open stage</a> or <a href='http://furinvader-feature-typing-unloaded-linked-resources.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com//teatro.log' target='_blank'>view logs</a>.

furinvader committed:
Make the  className parameter of halResourceType.add() optional

furinvader committed:
Refactor HalResource according to tslint

furinvader committed:
Refactor HalResource

This refactoring ensures that no 'private' methods
get exposed and possibly overridden.

furinvader committed:
Correctly initialize a HalResource after it has been loaded

furinvader committed:
Implement getType() in halResourceTypesStorage

furinvader committed:
Refactor tests

furinvader committed:
Refactor halResourceTypes and related services

Implement halResourceTypesStorage service
using a class instead of it being a value.

furinvader committed:
Implement better HalResource types configuration

furinvader committed:
Make $initialize work on already loaded resources

oliverguenther commented:
Taking a look at it now

oliverguenther commented:
LGTM 👍
